Small refactorings around partitioning.

hlinnaka opened the pull request:
No user visible effects. See commit messages.

ashwinstar commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2471,8 +2472,10 @@ cdbpathtoplan_create_sri_plan(RangeTblEntry *rte, PlannerInfo *subroot, Path *su
-		nulls = palloc0(sizeof(bool) * rel->rd_att->natts);
-		values = palloc(sizeof(Datum) * rel->rd_att->natts);
+		slot = MakeSingleTupleTableSlot(RelationGetDescr(rel));
+		ExecClearTuple(slot);
```
after `MakeSingleTupleTableSlot()`, `ExecClearTuple()` seems unnecessary?

hlinnaka replied:
Yeah, it probably is. I'd prefer to keep it, though. The comment in ExecStoreVirtualTuple says:
```
 * The protocol for loading a slot with virtual tuple data is:
 *		* Call ExecClearTuple to mark the slot empty.
 *		* Store data into the Datum/isnull arrays.
 *		* Call ExecStoreVirtualTuple to mark the slot valid.
 * This is a bit unclean but it avoids one round of data copying.
```
That explicitly says to call ExecClearTuple. Maybe it's not needed if the slot is empty already, but I'd rather play it safe.

ashwinstar commented on src/include/nodes/execnodes.h:
```diff
@@ -441,38 +441,13 @@ typedef struct ShareNodeEntry
  * PartitionMetadata
  *   Defines the metadata for partitions.
  */
 typedef struct PartitionMetadata
```
Should we get rid of this structure as well then and directly use `PartitionNode`?

hlinnaka replied:
Done

ashwinstar commented on src/include/executor/executor.h:
```diff
@@ -241,9 +241,8 @@ extern void EvalPlanQualEnd(EPQState *epqstate);
 extern Oid GetIntoRelOid(QueryDesc *queryDesc);
 
-extern bool map_part_attrs(Relation base, Relation part, TupleConversionMap **map_ptr, bool throwerror);
+extern bool map_part_attrs(Relation base, Relation part, TupleConversionMap **map_ptr, bool throw);
```
`throw` what :-) ?

hlinnaka replied:
Heh :-). I changed it to 'throw', because that's what the parameter is called in the actual function definition in `execMain.c`. But now I see why it was called 'throwerror' in the header:
```
In file included from ../../../../src/include/funcapi.h:22,
                 from ../../../../src/include/gpopt/utils/gpdbdefs.h:69,
                 from COptTasks.cpp:16:
../../../../src/include/executor/executor.h:244:93: error: expected ‘,’ or ‘...’ before ‘throw’
 extern bool map_part_attrs(Relation base, Relation part, TupleConversionMap **map_ptr, bool throw);
                                                                                             ^~~~~
make[1]: *** [<builtin>: COptTasks.o] Error 1
make: *** [../../../src/backend/common.mk:41: utils-recursive] Error 2
```
`throw` is a reserved keyword in C++, so it errors out when it's included in ORCA code that is compiled as C++.

Changed it back to 'throwerror'.

ashwinstar commented on src/include/nodes/execnodes.h:
```diff
@@ -441,38 +441,13 @@ typedef struct ShareNodeEntry
- * PartitionAccessMethods
- *    Defines the lookup access methods for partitions, one for each level.
- */
-typedef struct PartitionAccessMethods
```
This would have been much more confusing to read once AM api's land in GPDB.

hlinnaka commented:
Fixed some "tupledesc reference leak" warnings, a compiler warning/error.

Also fixed a failure in 'gporca' test that was caused by change in behaviour if you use an incomplete operator class with list partitioning. In PostgreSQL v10, you get an error at CREATE TABLE already, if you try to do that, so I don't much care for that case. Defining an incomplete operator class is really a user error to begin with.

hlinnaka commented:
I removed the fourth commit from this PR from now. That was a few bricks shy of a load: it cached the comparison functions the partition key's datatype, in the PartitionNode struct, but in some scenarios we needed the cross-datatype comparison functions instead. Something like that cleanup is still a good idea, but since it needs a bit more work, let's get these other commits pushed first.

ashwinstar commented on src/include/cdb/cdbpartition.h:
```diff
@@ -36,28 +36,15 @@ typedef enum
-	int last_rule; /* cache offset to the last rule and test if it matches */
-	PartitionRule **rules;
+	FmgrInfo *cmpfuncs;
 } PartitionRangeState;
```
Seems `PartitionRangeState` and `PartitionListState` both can use same struct as `PartitionState` or better would be to not even have struct for them now.

hlinnaka replied:
True. Good idea, I removed the structs, and just pass around FmgrInfo arrays now.

hlinnaka commented on src/backend/cdb/cdbutil.c:
```diff
@@ -674,7 +674,8 @@ cdbcomponent_updateCdbComponents(void)
 	PG_CATCH();
 	{
-		FtsNotifyProber();
+		if (Gp_session_role == GP_ROLE_DISPATCH)
```
Yeah, this is a bit out of place here...

What happened was that when running in pg_upgrade, we called getgpsegmentCount() to set the number of segments in a CdbPathLocus. However, everything actually runs in General locus in pg_upgrade / utility mode, so didn't really need the number of segments for it. https://github.com/greenplum-db/gpdb/pull/9353 will change that. After that PR, this change in cdbcomponent_updateCdbComponents() shouldn't be needed anymore.

It's not entirely clear to me why we needed that change with this PR, and not before. I think it was just because of the test case changes in this PR, we exercised some code in the pg_upgrade path that wasn't exercised before. In other words, I think that was an unrelated, pre-existing issue, but it wasn't exposed until the test case changes in this PR.

Anyway, let's see if this can go away after https://github.com/greenplum-db/gpdb/pull/9353 is pushed.

hlinnaka replied:
Ok, this is now rebased over PR #9353, so that this change is no longer needed.

hlinnaka replied:
I was way off with my original diagnosis. I was seeing a different error on my laptop than what the pipeline was hitting, because I had set up the test on my laptop differently.

I opened issue https://github.com/greenplum-db/gpdb/issues/9405 for the issue that I *think* this PR is hitting in concourse. I'm at a loss why we're not seeing it without this PR, though. This moves the test case from 'orca.sql' to 'partition.sql', but as far as I can see the old test should've hit the pg_upgrade issue, too.

hlinnaka replied:
Figured it out and fixed. There was another tiny silly bug in this PR that caused pg_upgrade to fail. The concourse output isn't great when the pg_upgrade step fails. It doesn't print the logs or the failed dump file. So I was confused on why it failed.
